messaging/src/main/java/org/axonframework/messaging/annotation/ResultParameterResolverFactory.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright (c) 2010-2020. Axon Framework
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.axonframework.messaging.annotation;
+
+import org.axonframework.common.annotation.AnnotationUtils;
+import org.axonframework.eventhandling.interceptors.ResultHandler;
+import org.axonframework.messaging.Message;
+
+import java.lang.reflect.Executable;
+import java.lang.reflect.Parameter;
+import java.util.concurrent.Callable;
+import java.util.function.Supplier;
+
+/**
+ * ParameterResolverFactory that provides support for Parameters where the result of Handler execution is expected to
+ * be injected. This is only possible in interceptor handlers that need to act on the result of downstream interceptors
+ * or the regular handler.
+ * <p>
+ * The {@link ResultHandler @ResultHandler} Meta-Annotation needs to be placed on handlers that support interacting with
+ * the result type in its parameters.
+ *
+ * @author Allard Buijze
+ * @since 4.4
+ */
+public class ResultParameterResolverFactory implements ParameterResolverFactory {
+
+
+    private static final ThreadLocal<Object> REGISTERED_RESULT = new ThreadLocal<>();
+    private static final Object IGNORE_RESULT_PARAMETER_MARKER = new Object();
+
+    /**
+     * Calls the given {@code action} (typically a handler invocation) such that the given {@code result} is available
+     * for injection as parameter
+     *
+     * @param result The result to make available for parameter injection
+     * @param action The action to take
+     *
+     * @return the result of the action
+     * @throws Exception any exception thrown while executing the {@code action}
+     */
+    public static Object callWithResult(Object result, Callable<?> action) throws Exception {
+        Object previous = REGISTERED_RESULT.get();
+        REGISTERED_RESULT.set(result);
+        try {
+            return action.call();
+        } finally {
+            if (previous == null) {
+                REGISTERED_RESULT.remove();
+            } else {
+                REGISTERED_RESULT.set(previous);
+            }
+        }
+    }
+
+    /**
+     * Performs the given {@code action} ignoring any parameters expecting a result type. This is typically used to
+     * detect whether a handler is suitable for invocation prior to the result value being available.
```

Comment:
So, the `canHandle` method will/can actually call `handle` too? Or, will the parameter already be resolved if we perform `canHandle`?

Reply:
This will not call `handle`. The idea is that we can use this method to find out if an interceptor acting on the result is _potentially_ able to handle it. But since we don't know the actual result of the invocation yet, we can't be sure. The actual _handle_ method is responsible for that additional check once the result is known.

Reply:
Had to debug this myself to truelly understand the process why this makes sense. Any how, it makes sense now, hence I'll resolve this comment too.